app/src/mouse/hid_input_listener.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/input/input.h>
+#include <zephyr/dt-bindings/input/input-event-codes.h>
+
+#include <zmk/mouse.h>
+#include <zmk/mouse/input_config.h>
+#include <zmk/endpoints.h>
+#include <zmk/hid.h>
+
+void handle_rel_code(struct input_event *evt) {
+    switch (evt->code) {
+    case INPUT_REL_X:
+        zmk_hid_mouse_movement_update(evt->value, 0);
+        break;
+    case INPUT_REL_Y:
+        zmk_hid_mouse_movement_update(0, evt->value);
+        break;
+    case INPUT_REL_WHEEL:
+        zmk_hid_mouse_scroll_update(0, evt->value);
+        break;
+    case INPUT_REL_HWHEEL:
+        zmk_hid_mouse_scroll_update(evt->value, 0);
+        break;
+    default:
+        break;
+    }
+}
+
+void handle_key_code(struct input_event *evt) {
+    int8_t btn;
+
+    switch (evt->code) {
+    case INPUT_BTN_0:
+    case INPUT_BTN_1:
+    case INPUT_BTN_2:
+    case INPUT_BTN_3:
+    case INPUT_BTN_4:
+        btn = evt->code - INPUT_BTN_0;
+        if (evt->value > 0) {
+            zmk_hid_mouse_button_press(btn);
+        } else {
+            zmk_hid_mouse_button_release(btn);
+        }
+        break;
+    default:
+        break;
+    }
+}
+
+static void swap_xy(struct input_event *evt) {
+    switch (evt->code) {
+    case INPUT_REL_X:
+        evt->code = INPUT_REL_Y;
+        break;
+    case INPUT_REL_Y:
+        evt->code = INPUT_REL_X;
+        break;
+    }
+}
+
+static void filter_with_input_config(struct input_event *evt) {
+    if (!evt->dev) {
+        return;
+    }
+
+    const struct zmk_input_config *cfg = zmk_input_config_get_for_device(evt->dev);
+
+    if (!cfg) {
+        return;
+    }
+
+    if (cfg->xy_swap) {
+        swap_xy(evt);
+    }
+
+    if ((cfg->x_invert && evt->code == INPUT_REL_X) ||
+        (cfg->y_invert && evt->code == INPUT_REL_Y)) {
+        evt->value = -(evt->value);
+    }
+
+    evt->value = (int16_t)((evt->value * cfg->scale_multiplier) / cfg->scale_divisor);
+}
+
+void input_handler(struct input_event *evt) {
+    // First, filter to update the event data as needed.
+    filter_with_input_config(evt);
+
+    switch (evt->type) {
+    case INPUT_EV_REL:
+        handle_rel_code(evt);
+        break;
+    case INPUT_EV_KEY:
+        handle_key_code(evt);
+        break;
+    }
+
+    if (evt->sync) {
+        zmk_endpoints_send_mouse_report();
+        zmk_hid_mouse_scroll_set(0, 0);
+        zmk_hid_mouse_movement_set(0, 0);
+    }
+}
+
+INPUT_CALLBACK_DEFINE(NULL, input_handler);
```

Comment:
If the handler requries syncronization between multiple event types you really want to bind the callback to a specific device. If you trigger the callback unfiltered, any other event from other devices with the sync bit set is going to to trigger the handler (the report in this case), potentially with inconsistent status.

For this use case I'd recommend splitting the callback, the key code can be handled unflitered just fine so it can be its own thing. For the movement and scroller you should have a callback per device and specific the handler as the first argument of `INPUT_CALLBACK_DEFINE`. I like to have a binding to instantiate as a child of the device generating the events and then use `DT_INST_PARENT` to find its handler here (see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/input/input_keymap.c#L105C46-L105C60 for an example), but you may as just have a known nodelabel and makes things even simpler.

Reply:
@fabiobaltieri Yeah, that makes a lot of sense! We probably should be staging the in-progress movement data per-device as well so that the sync flag can then be used to submit only that data to our HID layer.

That does mean we'll need to explicitly list all devices we want to have considered for input, instead of a non-filtered subscription as a catch all.

Reply:
Yeah that's the proper way of doing it, the Linux parallel of that would be opening a specific `/dev/input/event*` device node.

Reply:
@fabiobaltieri Ok, so I pushed a change here so that any input device you want to use requires you add an explicit listener node, e.g.:

```
    glidepoint_input {
        compatible = "zmk,input-listener";
        device = <&glidepoint>;

        xy-swap;
        y-invert;
        x-invert;
    };
```

That then creates a dedicated listener that buffers the event data scoped to that device until it gets a sync event, which causes the locally buffered data to be pushed into our HID layer.